### PR TITLE
feat(api): implement CS-028 registration flow

### LINKED ISSUE #48: [CS-028] As a new user I want to register with email and password so that I can access the platform
## Persona

- New user registering with local credentials

---

## Problem

New users need a secure, standards-compliant registration flow that creates a local account, issues a 24-hour verification token, and sends a verification email without blocking on provider failures.

---

## Scope Boundaries

- In scope:
  - Register endpoint with DTO validation and canonical error handling.
  - bcrypt hashing (cost 12) and email normalization.
  - Verification token generation/storage with 24h expiry (hashed, single-use).
  - Enqueue verification email job and worker dispatch with retry.
  - Register rate-limit keys (per IP + per email) aligned to auth guard config.
- Out of scope:
  - OAuth login flow (CS-031).
  - Account linking beyond explicit error codes.
  - Frontend registration UI enhancements.

---

## Architectural Context

- Auth is global; no workspace scoping.
- Registration creates `is_verified=false` users and does not issue sessions.
- Verification tokens are opaque, hashed, single-use, and expire in 24 hours.
- Email provider failures must not block registration; retries are handled via worker.
- Rate limiting uses the shared P17 RateLimitGuard with per-IP and per-email buckets.

---

## Acceptance Criteria

- [ ] AC1: Valid registration returns 201 message-only, emits `user.registered`, and enqueues a verification email; registration succeeds even if email delivery fails.
- [ ] AC2: Duplicate emails return `EMAIL_ALREADY_EXISTS` for local accounts and `ACCOUNT_EXISTS_OAUTH` for OAuth-owned emails using the canonical error envelope.
- [ ] AC3: Register is rate limited at 5/hour per IP and 5/hour per email, returning 429 `RATE_LIMITED` with `Retry-After`.

---

## Edge Cases

- [ ] Case-insensitive email duplicates are rejected consistently.
- [ ] Email provider failure still returns 201; verification email is retried asynchronously.
- [ ] Weak passwords return `VALIDATION_ERROR` or `PASSWORD_TOO_WEAK`.
- [ ] OAuth-owned email registration returns `ACCOUNT_EXISTS_OAUTH`.

---

## UI / Design

- Figma: N/A
- Key interactions: N/A
- Responsive: N/A

---

## API Contract

- Endpoints: `POST /api/v1/auth/register`
- Request/Response: `{ fullName, email, password }` → `201 { message }`
- Error codes: `VALIDATION_ERROR`, `EMAIL_ALREADY_EXISTS`, `ACCOUNT_EXISTS_OAUTH`, `RATE_LIMITED`

---

## Data / Migration

- DB change required: no
- Schema impact: `users`, `email_verification_tokens`
- Migration name: N/A
- Rollback plan: N/A

---

## Events / Realtime

- Domain events: `user.registered`, `user.verification_sent`
- Socket.IO events: none
- Collab (Hocuspocus/Yjs) impact: none

---

## Security / RBAC

- Required roles: any user
- Workspace scoping required: no
- Ownership checks: no
- Additional: bcrypt cost 12, token hashes only, 24h verification expiry, non-blocking email dispatch.

---

## Tests

- Unit: DTO validation, hashing, email normalization.
- Integration: register creates user + verification token; email job enqueued.
- E2E: register flow with mocked email dispatch.

---

## Observability

- Logs: registration attempts and email job enqueue outcomes.
- Metrics: register success/failure counts, rate-limit hits.
- Alerts: email worker failure alerts.

---

## Task Generation Guidance

- Task kernel candidates: register endpoint, hashing/normalization, verification token issuance, email job enqueue, email worker dispatch, rate-limit key wiring.
- ExecPlan triggers: new email provider integration or queue infrastructure changes.
- Expected routing per task: api, worker.
- Boundary rule: split by execution surface; avoid mixing API logic with worker dispatch or audit handling.
- Side effects: emit `user.registered` and `user.verification_sent` explicitly in deliverables.
- Validation: include negative-path checks for duplicate email, OAuth-owned email, weak password, and rate limits.
- Validation tasks: must include final story validation task.

---

## Planned Tasks

- [x] #347
- [x] #348
- [x] #349
- [x] #350
- [x] #351
- [x] #352
- [x] #1190

---

## Story Validation

- Required final task: CS-028-VALIDATE
- Validation focus: api-contract, integration, edge-cases, ops
- Validation labels: `type:validation`, `role:validate`, `spec:story-validation`, `gate:story-validation`

---

## Canonical References

- Primary: `docs/spec/04-user-flows/04.2-user-onboarding.md`, `docs/spec/05-features/05.1-authentication.md`, `docs/spec/08-data-model/08.3-auth-users.md`, `docs/spec/09-api-standards/09.7-auth-standards.md`, `docs/spec/09-api-standards/09.9-rate-limits.md`, `docs/spec/11-security/11.3-authentication-security.md`, `docs/spec/11-security/11.6-rate-limits-abuse-prevention.md`, `docs/spec/12-errors/12.4-error-code-catalog.md`, `docs/spec/18-appendices/18.1-domain-event-catalog.md`
- Secondary: `docs/agent-ref/api/auth-endpoints.md`, `docs/agent-ref/data/auth-schema.md`, `docs/agent-ref/rules/security-rules.md`, `docs/agent-ref/rules/error-codes.md`, `docs/agent-ref/rules/rate-limits.md`, `docs/agent-ref/events/domain-events.md`

---

## Execution References

- `AGENTS.md`
- `.agent/PLANS.md` (if any task is tier:S)
- `docs/agent-ref/*`

---

## Definition of Done

- [ ] All acceptance criteria met.
- [ ] Edge cases handled.
- [ ] Tests passing or explicitly waived with rationale.
- [ ] Story validation task completed with evidence.
- [ ] Docs updated if required.

### LINKED ISSUE #347: [CS-028-1] Implement POST /api/v1/auth/register endpoint with DTO validation and account-linking guard
## Objective

- Implement the register endpoint with DTO validation and explicit account-linking guards.

---

## Task Boundary Guidance

- Prefer **one primary execution surface** and **one dominant specialization** per task (e.g., API endpoint, UI view, worker job, integration test).
- Avoid mixing backend behavior + frontend UI + audit/telemetry in one task unless the change is tiny and inseparable.
- If a task must include a side effect (audit log, event emission), make it explicit in the objective and validation.
- Primary surface: API endpoint handler and request validation.
- Keep hashing, token creation, and worker dispatch in separate tasks.
- Account-linking guard logic belongs here; email sending does not.

---

## Deliverables

- POST `/api/v1/auth/register` endpoint validates input and returns 201 message-only response.
- Emits `user.registered` on successful registration.
- Returns `ACCOUNT_EXISTS_OAUTH` when email belongs to OAuth-owned account and `EMAIL_ALREADY_EXISTS` for local duplicates.

---

## Files

| Path | Change Type | Owner | Notes |
|---|---|---|---|
| apps/api/src/modules/auth/auth.controller.ts | edit | api | Register endpoint handler. |
| apps/api/src/modules/auth/auth.service.ts | edit | api | Registration flow logic. |
| apps/api/src/modules/auth/dto/register.dto.ts | edit | api | Request validation schema. |
| apps/api/src/modules/auth/auth.module.ts | edit | api | Module wiring. |

---

## Spec References

- `docs/agent-ref/api/auth-endpoints.md`
- `docs/agent-ref/data/auth-schema.md`
- `docs/agent-ref/rules/error-codes.md`
- `docs/agent-ref/rules/security-rules.md`
- `docs/spec/04-user-flows/04.2-user-onboarding.md`
- `docs/spec/05-features/05.1-authentication.md`
- `docs/spec/09-api-standards/09.7-auth-standards.md`
- `docs/spec/12-errors/12.4-error-code-catalog.md`

---

## Implementation Notes

- Ensure responses follow auth endpoint contract with canonical error envelope.
- Account linking: OAuth-owned email returns `ACCOUNT_EXISTS_OAUTH` (409).

---

## Constraints

- Workspace isolation: all workspace-owned queries must include `workspace_id`.
- REST must not return collaborative Yjs content.
- Follow response envelope and canonical error codes.
- Follow module AGENTS and `docs/agent-ref` contracts.
- Registration responses must not include access/refresh tokens.

---

## Validation

Commands:
```bash
pnpm -w lint
pnpm -w test
API_BASE_URL="${API_BASE_URL:-http://localhost:3001}"
curl -fsS -X POST "$API_BASE_URL/api/v1/auth/register" \
  -H "Content-Type: application/json" \
  -d '{"fullName":"Jane Doe","email":"jane+reg@example.com","password":"StrongPass@123"}'
```
Expected outcomes:
- Register returns 201 with message-only response and emits `user.registered`.
- Duplicate OAuth-owned email returns `ACCOUNT_EXISTS_OAUTH`.
- If not run, explain why and risk.

---

## Done When

- [ ] Deliverables implemented.
- [ ] Validation commands executed or explicitly waived.
- [ ] Errors/envelopes match `docs/agent-ref`.
- [ ] API tasks: endpoint contract and emitted events verified.
- [ ] No scope creep beyond listed files.
- [ ] Handoff completed.
- [ ] UI tasks: loading/empty/error/loaded states, accessibility, responsive behavior, and permission-gated states verified (if applicable).

---

## Handoff

- Summary of changes: pending
- Validation evidence: pending
- Open issues / follow-ups: pending
- Risks or deviations: pending

---

## References

- `AGENTS.md`
- `apps/api/AGENTS.md`
- `docs/agent-ref/*`

### LINKED ISSUE #348: [CS-028-2] Add bcrypt hashing (cost 12) and email normalization for registration
## Objective

- Add bcrypt password hashing (cost 12) and normalize emails before persistence.

---

## Task Boundary Guidance

- Prefer **one primary execution surface** and **one dominant specialization** per task (e.g., API endpoint, UI view, worker job, integration test).
- Avoid mixing backend behavior + frontend UI + audit/telemetry in one task unless the change is tiny and inseparable.
- If a task must include a side effect (audit log, event emission), make it explicit in the objective and validation.
- Primary surface: auth service logic for hashing and normalization.
- Do not add token issuance or email dispatch logic here.

---

## Deliverables

- Passwords are hashed with bcrypt cost 12 before persistence.
- Emails are normalized (lowercase + trim) before uniqueness checks and storage.

---

## Files

| Path | Change Type | Owner | Notes |
|---|---|---|---|
| apps/api/src/modules/auth/auth.service.ts | edit | api | Password hashing and normalization. |

---

## Spec References

- `docs/agent-ref/data/auth-schema.md`
- `docs/agent-ref/rules/security-rules.md`
- `docs/spec/04-user-flows/04.2-user-onboarding.md`
- `docs/spec/05-features/05.1-authentication.md`
- `docs/spec/11-security/11.3-authentication-security.md`

---

## Implementation Notes

- Normalize email early to ensure case-insensitive uniqueness.

---

## Constraints

- Workspace isolation: all workspace-owned queries must include `workspace_id`.
- REST must not return collaborative Yjs content.
- Follow response envelope and canonical error codes.
- Follow module AGENTS and `docs/agent-ref` contracts.
- Use bcrypt cost factor 12.

---

## Validation

Commands:
```bash
pnpm -w lint
pnpm -w test
```
Expected outcomes:
- Password hashes use bcrypt with cost 12.
- Normalized emails are persisted and compared case-insensitively.
- If not run, explain why and risk.

---

## Done When

- [ ] Deliverables implemented.
- [ ] Validation commands executed or explicitly waived.
- [ ] Errors/envelopes match `docs/agent-ref`.
- [ ] API tasks: endpoint contract and emitted events verified.
- [ ] No scope creep beyond listed files.
- [ ] Handoff completed.
- [ ] UI tasks: loading/empty/error/loaded states, accessibility, responsive behavior, and permission-gated states verified (if applicable).

---

## Handoff

- Summary of changes: pending
- Validation evidence: pending
- Open issues / follow-ups: pending
- Risks or deviations: pending

---

## References

- `AGENTS.md`
- `apps/api/AGENTS.md`
- `docs/agent-ref/*`

### LINKED ISSUE #349: [CS-028-3] Create email verification token generation/storage with 24-hour expiry
## Objective

- Generate and store email verification tokens with 24-hour expiry and hash-only storage.

---

## Task Boundary Guidance

- Prefer **one primary execution surface** and **one dominant specialization** per task (e.g., API endpoint, UI view, worker job, integration test).
- Avoid mixing backend behavior + frontend UI + audit/telemetry in one task unless the change is tiny and inseparable.
- If a task must include a side effect (audit log, event emission), make it explicit in the objective and validation.
- Primary surface: token generation and persistence logic.
- Do not enqueue email jobs or handle worker dispatch here.

---

## Deliverables

- Verification tokens are generated as opaque values, hashed with sha256, and stored with 24-hour expiry.
- Tokens are single-use and ready for verification flows.

---

## Files

| Path | Change Type | Owner | Notes |
|---|---|---|---|
| apps/api/src/modules/auth/auth.service.ts | edit | api | Verification token lifecycle. |
| apps/api/src/modules/auth/tokens.service.ts | edit | api | Token helpers (file name may vary). |

---

## Spec References

- `docs/agent-ref/data/auth-schema.md`
- `docs/agent-ref/rules/security-rules.md`
- `docs/spec/04-user-flows/04.2-user-onboarding.md`
- `docs/spec/05-features/05.1-authentication.md`
- `docs/spec/11-security/11.3-authentication-security.md`

---

## Implementation Notes

- Store only token hashes and expires_at; never store plaintext tokens.

---

## Constraints

- Workspace isolation: all workspace-owned queries must include `workspace_id`.
- REST must not return collaborative Yjs content.
- Follow response envelope and canonical error codes.
- Follow module AGENTS and `docs/agent-ref` contracts.
- Verification token expiry must be exactly 24 hours.

---

## Validation

Commands:
```bash
pnpm -w lint
pnpm -w test
```
Expected outcomes:
- Token hashes are stored with 24-hour expiry.
- Tokens are single-use and not stored in plaintext.
- If not run, explain why and risk.

---

## Done When

- [ ] Deliverables implemented.
- [ ] Validation commands executed or explicitly waived.
- [ ] Errors/envelopes match `docs/agent-ref`.
- [ ] API tasks: endpoint contract and emitted events verified.
- [ ] No scope creep beyond listed files.
- [ ] Handoff completed.
- [ ] UI tasks: loading/empty/error/loaded states, accessibility, responsive behavior, and permission-gated states verified (if applicable).

---

## Handoff

- Summary of changes: pending
- Validation evidence: pending
- Open issues / follow-ups: pending
- Risks or deviations: pending

---

## References

- `AGENTS.md`
- `apps/api/AGENTS.md`
- `docs/agent-ref/*`

### LINKED ISSUE #350: [CS-028-4] Enqueue verification email job on registration success
## Objective

- Enqueue a verification email job after successful registration without blocking the API response.

---

## Task Boundary Guidance

- Prefer **one primary execution surface** and **one dominant specialization** per task (e.g., API endpoint, UI view, worker job, integration test).
- Avoid mixing backend behavior + frontend UI + audit/telemetry in one task unless the change is tiny and inseparable.
- If a task must include a side effect (audit log, event emission), make it explicit in the objective and validation.
- Primary surface: API enqueue logic only.
- Worker job implementation and provider integration are handled in CS-028-6.

---

## Deliverables

- Registration flow enqueues a verification email job with token reference.
- Registration returns 201 even if the provider is unavailable.

---

## Files

| Path | Change Type | Owner | Notes |
|---|---|---|---|
| apps/api/src/modules/auth/auth.service.ts | edit | api | Queue email dispatch. |

---

## Spec References

- `docs/agent-ref/api/auth-endpoints.md`
- `docs/agent-ref/rules/security-rules.md`
- `docs/spec/04-user-flows/04.2-user-onboarding.md`
- `docs/spec/05-features/05.1-authentication.md`

---

## Implementation Notes

- Do not attempt provider delivery in the API layer; only enqueue.

---

## Constraints

- Workspace isolation: all workspace-owned queries must include `workspace_id`.
- REST must not return collaborative Yjs content.
- Follow response envelope and canonical error codes.
- Follow module AGENTS and `docs/agent-ref` contracts.
- Registration responses must not include access/refresh tokens.

---

## Validation

Commands:
```bash
pnpm -w lint
pnpm -w test
```
Expected outcomes:
- Register enqueues a verification email job without blocking the response.
- Provider failures do not change the 201 response.
- If not run, explain why and risk.

---

## Done When

- [ ] Deliverables implemented.
- [ ] Validation commands executed or explicitly waived.
- [ ] Errors/envelopes match `docs/agent-ref`.
- [ ] API tasks: endpoint contract and emitted events verified.
- [ ] No scope creep beyond listed files.
- [ ] Handoff completed.
- [ ] UI tasks: loading/empty/error/loaded states, accessibility, responsive behavior, and permission-gated states verified (if applicable).

---

## Handoff

- Summary of changes: pending
- Validation evidence: pending
- Open issues / follow-ups: pending
- Risks or deviations: pending

---

## References

- `AGENTS.md`
- `apps/api/AGENTS.md`
- `docs/agent-ref/*`

### LINKED ISSUE #351: [CS-028-5] Apply register rate limit keys (per IP + per email) using RateLimitGuard config
## Objective

- Ensure register endpoint rate limits use per-IP and per-email keys with canonical thresholds.

---

## Task Boundary Guidance

- Prefer **one primary execution surface** and **one dominant specialization** per task (e.g., API endpoint, UI view, worker job, integration test).
- Avoid mixing backend behavior + frontend UI + audit/telemetry in one task unless the change is tiny and inseparable.
- If a task must include a side effect (audit log, event emission), make it explicit in the objective and validation.
- Primary surface: endpoint guard metadata and key selection.
- Do not modify rate limit infrastructure owned by P17.

---

## Deliverables

- Register endpoint uses RateLimitGuard keys for IP + email buckets.
- 429 responses include `Retry-After` and `RATE_LIMITED`.

---

## Files

| Path | Change Type | Owner | Notes |
|---|---|---|---|
| apps/api/src/modules/auth/auth.controller.ts | edit | api | Apply guard metadata for register. |
| apps/api/src/modules/auth/auth.module.ts | edit | api | Ensure guard is wired to register route. |

---

## Spec References

- `docs/agent-ref/rules/rate-limits.md`
- `docs/agent-ref/rules/security-rules.md`
- `docs/spec/05-features/05.1-authentication.md`
- `docs/spec/09-api-standards/09.9-rate-limits.md`
- `docs/spec/11-security/11.6-rate-limits-abuse-prevention.md`

---

## Implementation Notes

- Use CS-038 configuration for thresholds (5/hour per IP + 5/hour per email).

---

## Constraints

- Workspace isolation: all workspace-owned queries must include `workspace_id`.
- REST must not return collaborative Yjs content.
- Follow response envelope and canonical error codes.
- Follow module AGENTS and `docs/agent-ref` contracts.
- Rate limiting must not leak account existence.

---

## Validation

Commands:
```bash
pnpm -w lint
pnpm -w test
API_BASE_URL="${API_BASE_URL:-http://localhost:3001}"
curl -fsS -X POST "$API_BASE_URL/api/v1/auth/register" \
  -H "Content-Type: application/json" \
  -d '{"fullName":"Jane Doe","email":"rate-limit-test@example.com","password":"StrongPass@123"}'
```
Expected outcomes:
- Register returns 429 `RATE_LIMITED` after threshold with `Retry-After`.
- Non-enumerating behavior is preserved.
- If not run, explain why and risk.

---

## Done When

- [ ] Deliverables implemented.
- [ ] Validation commands executed or explicitly waived.
- [ ] Errors/envelopes match `docs/agent-ref`.
- [ ] API tasks: endpoint contract and emitted events verified.
- [ ] No scope creep beyond listed files.
- [ ] Handoff completed.
- [ ] UI tasks: loading/empty/error/loaded states, accessibility, responsive behavior, and permission-gated states verified (if applicable).

---

## Handoff

- Summary of changes: pending
- Validation evidence: pending
- Open issues / follow-ups: pending
- Risks or deviations: pending

---

## References

- `AGENTS.md`
- `apps/api/AGENTS.md`
- `docs/agent-ref/*`

### LINKED ISSUE #352: [CS-028-6] Implement worker job to send verification email and retry on provider failure
## Objective

- Send verification emails via worker job with retries and emit `user.verification_sent` on success.

---

## Task Boundary Guidance

- Prefer **one primary execution surface** and **one dominant specialization** per task (e.g., API endpoint, UI view, worker job, integration test).
- Avoid mixing backend behavior + frontend UI + audit/telemetry in one task unless the change is tiny and inseparable.
- If a task must include a side effect (audit log, event emission), make it explicit in the objective and validation.
- Primary surface: worker job execution and retry handling.
- Do not modify API endpoint behavior or response envelopes here.

---

## Deliverables

- Worker job sends verification email using provider integration.
- Emits `user.verification_sent` when email dispatch succeeds.
- Provider failures are logged and retried without affecting API responses.

---

## Files

| Path | Change Type | Owner | Notes |
|---|---|---|---|
| apps/worker/src/jobs/email/* | edit | worker | Verification email job implementation (exact filenames may vary). |

---

## Spec References

- `docs/agent-ref/events/domain-events.md`
- `docs/agent-ref/rules/security-rules.md`
- `docs/spec/04-user-flows/04.2-user-onboarding.md`
- `docs/spec/05-features/05.1-authentication.md`
- `docs/spec/11-security/11.3-authentication-security.md`

---

## Implementation Notes

- Ensure token values are not logged.
- Retries should follow existing worker retry/backoff conventions.

---

## Constraints

- Workspace isolation: all workspace-owned queries must include `workspace_id`.
- REST must not return collaborative Yjs content.
- Follow response envelope and canonical error codes.
- Follow module AGENTS and `docs/agent-ref` contracts.
- Do not surface provider failures to end users.

---

## Validation

Commands:
```bash
pnpm -w test
```
Expected outcomes:
- Verification email job sends successfully and retries on provider failure.
- `user.verification_sent` is emitted on successful send.
- If not run, explain why and risk.

---

## Done When

- [ ] Deliverables implemented.
- [ ] Validation commands executed or explicitly waived.
- [ ] Errors/envelopes match `docs/agent-ref`.
- [ ] API tasks: endpoint contract and emitted events verified.
- [ ] No scope creep beyond listed files.
- [ ] Handoff completed.
- [ ] UI tasks: loading/empty/error/loaded states, accessibility, responsive behavior, and permission-gated states verified (if applicable).

---

## Handoff

- Summary of changes: pending
- Validation evidence: pending
- Open issues / follow-ups: pending
- Risks or deviations: pending

---

## References

- `AGENTS.md`
- `apps/worker/AGENTS.md`
- `docs/agent-ref/*`

### PR DESCRIPTION
**Pull request by Rick1330**
## Linked Issue

Closes #48
Closes #347
Closes #348
Closes #349
Closes #350
Closes #351
Closes #352

## Summary

- Implemented `POST /api/v1/auth/register` in API bootstrap with JSON body parsing, payload validation, email normalization, conflict handling, and canonical response envelope behavior.
- Added registration service/repository flow with bcrypt (`cost=12`) password hashing, verification token generation + SHA-256 persistence, and 24-hour token expiry.
- Added register rate limiting (`5/hour` per IP and per normalized email) with canonical `RATE_LIMITED` handling and `Retry-After` header propagation through the error envelope pipeline.
- Added non-blocking verification-email queue handoff from API and worker-side queue processing with retry/backoff handling plus `user.registered` / `user.verification_sent` domain-event append surfaces.
- Added CS-028 unit coverage for hash/token behavior, conflict mapping, password policy, and register rate limiting behavior.

## Validation

- [x] `pnpm lint`
- [x] `pnpm --filter @collabsphere/api run test`
- [x] `pnpm --filter @collabsphere/worker run build`
- [x] `pnpm test:unit`
- [ ] `pnpm review:coderabbit` (local CLI could not complete scoped review due upstream service failure/limit response; using PR reviewers for full review pass)

## Risks / Notes

- CS-028 task templates target full NestJS module paths, but current runtime baseline is bootstrap service surfaces; implementation follows current live architecture while preserving canonical auth/error/rate-limit contracts.
- Worker email dispatch currently uses the bootstrap queue/event simulation surface and should be wired to the production provider adapter as that lane is finalized.

## Handoff

- [ ] Handoff not required for this PR
- [x] Handoff added to the linked issue or parent story

## Handoff Summary

- Delivered CS-028 registration stack end-to-end across API + worker bootstrap lanes: register endpoint, secure credential handling, verification token lifecycle, queue dispatch, retry processing, and canonical event/error surfaces.

## Handoff Validation Evidence

- `pnpm lint` passed.
- `pnpm --filter @collabsphere/api run test` passed.
- `pnpm --filter @collabsphere/worker run build` passed.
- `pnpm test:unit` passed.

## Handoff Open Issues / Follow-ups

- Story validation gate issue `#1190` remains open for validation-phase closure after this PR merges.
- Local CodeRabbit pre-PR run was attempted but failed upstream; PR reviewers are required to complete final review coverage.

## Handoff Risks / Deviations

- Local CodeRabbit CLI review could not complete because the service reported an oversized/failed review context; no local CodeRabbit findings were available to action in this pass.

## Review Guidance

- Relevant spec / agent-ref docs: `docs/spec/05-features/05.1-authentication.md`, `docs/spec/09-api-standards/09.7-auth-standards.md`, `docs/spec/09-api-standards/09.9-rate-limits.md`, `docs/spec/11-security/11.3-authentication-security.md`, `docs/spec/11-security/11.6-rate-limits-abuse-prevention.md`, `docs/spec/12-errors/12.4-error-code-catalog.md`, `docs/spec/18-appendices/18.1-domain-event-catalog.md`.
- Areas that need extra scrutiny: auth conflict-code mapping and error envelope alignment, rate-limit retry-header behavior, token handling at API->worker boundary, and worker retry/backoff behavior under simulated provider failure.

---

<a href="https://app.devin.ai/review/rick1330/collabsphere/pull/1590" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: apps/worker/src/dev.ts [hunk 1]
```diff
@@ -42,6 +43,10 @@ validateServiceEnv({
   parser: parseRuntimeEnv,
   validationErrorClass: EnvValidationError,
 });
+const stopVerificationEmailProcessor = startVerificationEmailProcessor({
+  jwtAccessSecret: process.env.JWT_ACCESS_SECRET ?? "",
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/worker/src/dev.ts]: Fixed in cae7668. Worker now stores the validated env result and passes `workerEnv.JWT_ACCESS_SECRET` and `workerEnv.BASE_URL` directly into the verification processor (no empty-string fallback).

File: apps/api/src/auth/auth.controller.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    const payload = validateRegisterInput(rawPayload);
+
+    try {
+      return registerService.register({
+        payload,
+        ipAddress: getRequestContext()?.ip ?? request.socket.remoteAddress ?? "unknown",
+      });
+    } catch (error) {
+      throw mapRegisterPersistenceError(error);
+    }
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/auth/auth.controller.ts:96]: Addressed in cae7668 with the same controller change: `return await registerService.register(...)` is now awaited inside try/catch for consistent error mapping.

File: apps/api/src/auth/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      passwordHash,
+    });
+    const verificationToken = createOpaqueToken();
+    const verificationTokenHash = hashSha256(verificationToken);
+    const expiresAt = new Date(now().getTime() + verificationTokenTtlMs);
+    const storedToken = await repository.createEmailVerificationToken({
+      userId: registeredUser.id,
+      tokenHash: verificationTokenHash,
+      expiresAt,
+    });
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/auth/auth.service.ts:190]: Fixed in cae7668 by making user creation + verification token creation atomic in `createLocalUserWithVerificationToken` using a Prisma transaction.

File: apps/api/src/common/filters/app-error.filter.ts [hunk 4]
```diff
@@ -362,6 +369,7 @@ export const createErrorResponse = ({
 
   return {
     statusCode: normalizedError.statusCode,
+    headers: normalizedError.headers,
     normalizedError,
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/common/filters/app-error.filter.ts:409]: Implemented in cae7668. `normalizeAppError` now sanitizes response headers through an allowlist and only forwards `Retry-After` for `RATE_LIMITED`.

File: apps/api/src/auth/auth.controller.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/auth/auth.controller.ts:96]: Fixed in cae7668: controller now uses `return await registerService.register(...)` inside the try/catch, so async persistence errors are mapped through `mapRegisterPersistenceError`.

File: apps/api/src/common/filters/app-error.filter.ts [hunk 5]
```diff
@@ -326,18 +331,20 @@ const normalizeAppError = (error: unknown): NormalizedAppError => {
...
-    statusCode: errorStatusByCode[code],
-    originalError: error,
-  };
+    return {
+      code,
+      message: getDefaultErrorMessage(code),
+      statusCode: errorStatusByCode[code],
+      headers: undefined,
+      originalError: error,
+    };
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/common/filters/app-error.filter.ts]: Fixed in cae7668. The `normalizeAppError` fallback return is now aligned with `const code` and keeps standard formatting.

File: apps/api/src/auth/register-rate-limit.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+type RateLimitBucket = {
+  limit: number;
+  windowMs: number;
+  timestamps: number[];
+};
+
+const registerLimitPerHour = 5;
+const oneHourMs = 60 * 60 * 1000;
+const secondsPerMillisecond = 1000;
+
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/auth/register-rate-limit.ts:12]: Updated in cae7668: renamed `secondsPerMillisecond` to `millisecondsPerSecond` for clarity in retry-after calculation.

File: apps/api/src/dev.ts [hunk 7]
```diff
@@ -36,6 +42,38 @@ try {
...
+  }
+};
+
+const getRegisterService = () => {
+  if (!registerServicePromise) {
+    registerServicePromise = createRuntimeRegisterService();
+  }
+
+  return registerServicePromise;
+};
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/dev.ts:76]: Fixed in cae7668. `getRegisterService()` now resets the cached promise to `null` when `createRuntimeRegisterService()` rejects, so retries can recover after transient startup failures.

File: apps/worker/src/jobs/email/verification-email-queue.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    pendingJobs.push(job);
+  }
+
+  if (dueJobs.length > 0) {
+    await writeJobs({
+      queueFilePath,
+      jobs: pendingJobs,
+    });
+  }
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/worker/src/jobs/email/verification-email-queue.ts]: Addressed in cae7668. Worker queue enqueue/dequeue now run under the same file lock strategy and atomic writes, preventing cross-process dropped jobs during concurrent producer/consumer operations.

File: apps/api/src/auth/verification-email-queue.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+export const enqueueVerificationEmailJob = async ({
+  job,
+  queueFilePath = defaultQueueFilePath(),
+}: {
+  job: VerificationEmailJob;
+  queueFilePath?: string;
+}) => {
+  const jobs = await readJobs({ queueFilePath });
+  jobs.push(job);
+  await writeJobs({ queueFilePath, jobs });
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply apps/api/src/auth/verification-email-queue.ts]: Addressed in cae7668. API queue writes are now guarded by a lock file (`.lock`) and use atomic temp-file rename writes to avoid enqueue/dequeue read-modify-write races.

File: tests/unit/api-auth-register.test.ts [hunk 10]
```diff
@@ -0,0 +1,226 @@
+import assert from "node:assert/strict";
+import test from "node:test";
+import { AuthProvider } from "@prisma/client";
```
[the remark this answers is not in the document]
Comment (Rick1330) [reply tests/unit/api-auth-register.test.ts]: Current branch no longer imports `@prisma/client` in `tests/unit/api-auth-register.test.ts`; the test uses only `RegisterRepository` types and passes under both `pnpm --filter @collabsphere/api run test` and `pnpm test:unit` in cae7668.

**Comment by Rick1330** [comment]
Implemented reviewer follow-ups in commit cae7668 and revalidated locally.

What changed:
- Fixed async error mapping in register controller (wait in try/catch).
- Made user + verification token persistence atomic via repository transaction.
- Restricted forwarded error headers to allowlisted Retry-After for RATE_LIMITED.
- Used validated worker env output directly for JWT secret/base URL.
- Added queue locking + atomic file writes for API enqueue and worker enqueue/dequeue to prevent concurrent RMW races.
- Fixed cached rejected register-service promise in API bootstrap by resetting cache on failure.
- Updated unit test repository double to match transactional repository contract.
- Minor naming cleanup in rate limit helper.

Local validation run on this commit:
- pnpm lint
- pnpm --filter @collabsphere/api run test
- pnpm --filter @collabsphere/worker run build
- pnpm test:unit

All passed locally.

File: apps/worker/src/jobs/email/verification-email-queue.ts [hunk 11]
```diff
@@ -0,0 +1,249 @@
+import { createDecipheriv, createHash, randomBytes } from "node:crypto";
```
Comment (Rick1330) [inline apps/worker/src/jobs/email/verification-email-queue.ts]: Acknowledged. The worker queue now mirrors the API-side lock/atomic-write hardening, which raised overall complexity score. I kept this in the same PR to close the concurrency gap; a follow-up refactor can consolidate queue primitives to lower complexity while preserving correctness.

File: apps/api/src/auth/verification-email-queue.ts [hunk 12]
```diff
@@ -0,0 +1,200 @@
+import { createCipheriv, createHash, randomBytes } from "node:crypto";
```
Comment (Rick1330) [inline apps/api/src/auth/verification-email-queue.ts]: Acknowledged. This increase came from adding explicit lock/atomic-write safeguards to prevent producer/consumer race conditions in file-backed queue I/O. I prefer the safety fix for this PR; I can follow up with a shared queue helper extraction to reduce module-level complexity without removing the race protection.

**Comment by Rick1330** [comment]
Status update (2026-04-25):

- All actionable Devin/Copilot inline findings have been addressed in cae7668 and replied inline.
- Local validation passed: pnpm lint, pnpm --filter @collabsphere/api run test, pnpm --filter @collabsphere/worker run build, pnpm test:unit.
- CI build/test checks are green.
- Remaining red checks are non-blocking CodeScene/Sonar quality gates (advisory/code-health + hotspots/duplication).

This PR is ready for elevated human review when available.
